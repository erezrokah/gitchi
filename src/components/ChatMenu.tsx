import * as React from 'react';
import { useState } from 'react';
import { Menu, MenuItemProps, Icon } from 'semantic-ui-react';

export const ChatMenu = () => {
  const [active, setActive] = useState('channel-1');

  const handleItemClick = (
    e: React.MouseEvent<HTMLAnchorElement>,
    { name }: MenuItemProps,
  ) => {
    setActive(name || '');
  };

  const handleClose = () => {};

  return (
    <Menu tabular attached="top">
      <Menu.Item
        name="channel-1"
        active={active === 'channel-1'}
        onClick={handleItemClick}
      >
        Channel 1
      </Menu.Item>
      <Menu.Item
        name="new-tab"
        active={active === 'new-tab'}
        onClick={handleClose}
        position="right"
      >
        <Icon name="close" />
        Close
      </Menu.Item>
    </Menu>
  );
};
