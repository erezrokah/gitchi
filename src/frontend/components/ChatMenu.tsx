import * as React from 'react';
import { useState, useEffect } from 'react';
import { Menu, MenuItemProps, Icon } from 'semantic-ui-react';
import styled from 'styled-components';

const StyledItem = styled(Menu.Item)`
  width: 130px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-align: center;
  display: block !important;
`;

interface Props {
  onMenuToggle: (collapsed: boolean) => void;
  onRefreshClicked: () => void;
  onChannelSelect: (key: string) => void;
  channels: Channel[];
}

export const ChatMenu = (props: Props) => {
  const { onMenuToggle, onChannelSelect, onRefreshClicked } = props;

  const [active, setActive] = useState(props.channels[0].key);
  const [collapsed, setCollapsed] = useState(false);

  const toggleCollapse = () => {
    setCollapsed(!collapsed);
    onMenuToggle(!collapsed);
  };

  const handleItemClick = (
    e: React.MouseEvent<HTMLAnchorElement>,
    { name }: MenuItemProps,
  ) => {
    setActive(name || '');
    onChannelSelect(name || '');

    if (collapsed) {
      toggleCollapse();
    }
  };

  // reset active if channel is removed
  useEffect(() => {
    if (!props.channels.some(c => c.key === active)) {
      setActive(props.channels[0].key);
    }
    return () => {};
  }, [props.channels, active]);

  const items = props.channels.map(({ key, title }) => {
    return (
      <StyledItem
        key={key}
        name={key}
        active={active === key}
        onClick={handleItemClick}
        title={title}
      >
        {title}
      </StyledItem>
    );
  });

  const collapseButton = collapsed ? (
    <Menu.Item
      name="open"
      active={false}
      position="right"
      onClick={toggleCollapse}
      title="Open"
    >
      <Icon name="arrow up" />
      Open
    </Menu.Item>
  ) : (
    <Menu.Item
      name="collapse"
      active={false}
      position="right"
      onClick={toggleCollapse}
      title="Collapse"
    >
      <Icon name="arrow down" />
      Collapse
    </Menu.Item>
  );

  const menus: React.ReactElement[] = [];
  const itemsPerRow = 4;
  for (let index = 0; index < items.length; index = index + itemsPerRow) {
    const part = items.slice(index, index + itemsPerRow);

    menus.push(
      <Menu tabular attached="top" key={index}>
        {part}
        {index + itemsPerRow >= items.length ? (
          <Menu.Item
            name="refresh"
            active={false}
            onClick={onRefreshClicked}
            title="Refresh"
          >
            <Icon name="refresh" />
          </Menu.Item>
        ) : null}
        {index === 0 ? collapseButton : null}
      </Menu>,
    );
  }

  return <div>{menus}</div>;
};
