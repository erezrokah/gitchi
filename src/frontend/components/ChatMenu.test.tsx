import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ChatMenu } from './ChatMenu';

describe('ChatMenu', () => {
  const channels = [
    { key: 'Main', title: 'Main', isReview: false, comments: [] },
    { key: 'Channel1', title: 'Channel1', isReview: true, comments: [] },
    { key: 'Channel2', title: 'Channel2', isReview: true, comments: [] },
    { key: 'Channel3', title: 'Channel3', isReview: true, comments: [] },
    { key: 'Channel4', title: 'Channel4', isReview: true, comments: [] },
  ];
  const props = {
    onMenuToggle: jest.fn(),
    onRefreshClicked: jest.fn(),
    onChannelSelect: jest.fn(),
    channels,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should match snapshot', () => {
    const { asFragment } = render(<ChatMenu {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should choose first channel as active by default', () => {
    const { getByTitle } = render(<ChatMenu {...props} />);
    expect(getByTitle(channels[0].title)).toHaveClass('active');
  });

  test('should be open by default', () => {
    const { getByTitle, queryByTitle } = render(<ChatMenu {...props} />);
    expect(getByTitle('Collapse')).toBeInTheDocument();
    expect(queryByTitle('Open')).toBeNull();
  });

  test('should collapse/open menu on button clicks', () => {
    const { getByTitle, queryByTitle, asFragment } = render(
      <ChatMenu {...props} />,
    );

    fireEvent.click(getByTitle('Collapse'));

    expect(props.onMenuToggle).toHaveBeenCalledTimes(1);
    expect(props.onMenuToggle).toHaveBeenCalledWith(true);
    expect(getByTitle('Open')).toBeInTheDocument();
    expect(queryByTitle('Collapse')).toBeNull();

    expect(asFragment()).toMatchSnapshot();

    jest.clearAllMocks();
    fireEvent.click(getByTitle('Open'));

    expect(props.onMenuToggle).toHaveBeenCalledTimes(1);
    expect(props.onMenuToggle).toHaveBeenCalledWith(false);
    expect(getByTitle('Collapse')).toBeInTheDocument();
    expect(queryByTitle('Open')).toBeNull();
  });

  test('should change active menu on menu item click', () => {
    const { getByTitle } = render(<ChatMenu {...props} />);

    fireEvent.click(getByTitle(channels[1].title));
    expect(getByTitle(channels[1].title)).toHaveClass('active');

    expect(props.onChannelSelect).toHaveBeenCalledTimes(1);
    expect(props.onChannelSelect).toHaveBeenCalledWith(channels[1].key);
    expect(props.onMenuToggle).toHaveBeenCalledTimes(0);
  });

  test('should change active menu on menu item click', () => {
    const { getByTitle } = render(<ChatMenu {...props} />);

    fireEvent.click(getByTitle(channels[1].title));

    expect(getByTitle(channels[1].title)).toHaveClass('active');
    expect(props.onChannelSelect).toHaveBeenCalledTimes(1);
    expect(props.onChannelSelect).toHaveBeenCalledWith(channels[1].key);
  });

  test('should open collapsed menu on menu item click', () => {
    const { getByTitle, asFragment } = render(<ChatMenu {...props} />);

    fireEvent.click(getByTitle('Collapse'));
    jest.clearAllMocks();

    fireEvent.click(getByTitle(channels[0].title));

    expect(props.onMenuToggle).toHaveBeenCalledTimes(1);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should not call onMenuToggle when clicking a menu item if menu is open', () => {
    const { getByTitle } = render(<ChatMenu {...props} />);

    fireEvent.click(getByTitle(channels[0].title));

    expect(props.onMenuToggle).toHaveBeenCalledTimes(0);
  });

  test('should reset active menu when active channel is removed', () => {
    const { getByTitle, rerender } = render(<ChatMenu {...props} />);

    fireEvent.click(getByTitle(props.channels[1].title));
    expect(getByTitle(props.channels[1].title)).toHaveClass('active');

    const channels = [...props.channels];
    channels.splice(1, 1);

    rerender(<ChatMenu {...props} channels={channels} />);

    expect(getByTitle(props.channels[0].title)).toHaveClass('active');
  });

  test('should not reset active menu when non active channel is removed', () => {
    const { getByTitle, rerender } = render(<ChatMenu {...props} />);

    fireEvent.click(getByTitle(props.channels[1].title));
    expect(getByTitle(props.channels[1].title)).toHaveClass('active');

    const channels = [...props.channels];
    channels.splice(2, 1);

    rerender(<ChatMenu {...props} channels={channels} />);

    expect(getByTitle(props.channels[1].title)).toHaveClass('active');
  });

  test('should call onRefreshClicked on refresh button click', () => {
    const { getByTitle } = render(<ChatMenu {...props} />);

    fireEvent.click(getByTitle('Refresh'));

    expect(props.onRefreshClicked).toHaveBeenCalledTimes(1);
  });
});
