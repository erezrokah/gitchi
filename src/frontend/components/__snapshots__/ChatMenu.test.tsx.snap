// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMenu should collapse/open menu on button clicks 1`] = `
<DocumentFragment>
  .c0 {
  width: 130px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-align: center;
  display: block !important;
}

<div>
    <div
      class="ui top attached tabular menu"
    >
      <a
        class="active item c0"
        title="Main"
      >
        Main
      </a>
      <a
        class="item c0"
        title="Channel1"
      >
        Channel1
      </a>
      <a
        class="item c0"
        title="Channel2"
      >
        Channel2
      </a>
      <a
        class="item c0"
        title="Channel3"
      >
        Channel3
      </a>
      <a
        class="right item"
        title="Open"
      >
        <i
          aria-hidden="true"
          class="arrow up icon"
        />
        Open
      </a>
    </div>
    <div
      class="ui top attached tabular menu"
    >
      <a
        class="item c0"
        title="Channel4"
      >
        Channel4
      </a>
      <a
        class="item"
        title="Refresh"
      >
        <i
          aria-hidden="true"
          class="refresh icon"
        />
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ChatMenu should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  width: 130px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-align: center;
  display: block !important;
}

<div>
    <div
      class="ui top attached tabular menu"
    >
      <a
        class="active item c0"
        title="Main"
      >
        Main
      </a>
      <a
        class="item c0"
        title="Channel1"
      >
        Channel1
      </a>
      <a
        class="item c0"
        title="Channel2"
      >
        Channel2
      </a>
      <a
        class="item c0"
        title="Channel3"
      >
        Channel3
      </a>
      <a
        class="right item"
        title="Collapse"
      >
        <i
          aria-hidden="true"
          class="arrow down icon"
        />
        Collapse
      </a>
    </div>
    <div
      class="ui top attached tabular menu"
    >
      <a
        class="item c0"
        title="Channel4"
      >
        Channel4
      </a>
      <a
        class="item"
        title="Refresh"
      >
        <i
          aria-hidden="true"
          class="refresh icon"
        />
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ChatMenu should open collapsed menu on menu item click 1`] = `
<DocumentFragment>
  .c0 {
  width: 130px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-align: center;
  display: block !important;
}

<div>
    <div
      class="ui top attached tabular menu"
    >
      <a
        class="active item c0"
        title="Main"
      >
        Main
      </a>
      <a
        class="item c0"
        title="Channel1"
      >
        Channel1
      </a>
      <a
        class="item c0"
        title="Channel2"
      >
        Channel2
      </a>
      <a
        class="item c0"
        title="Channel3"
      >
        Channel3
      </a>
      <a
        class="right item"
        title="Collapse"
      >
        <i
          aria-hidden="true"
          class="arrow down icon"
        />
        Collapse
      </a>
    </div>
    <div
      class="ui top attached tabular menu"
    >
      <a
        class="item c0"
        title="Channel4"
      >
        Channel4
      </a>
      <a
        class="item"
        title="Refresh"
      >
        <i
          aria-hidden="true"
          class="refresh icon"
        />
      </a>
    </div>
  </div>
</DocumentFragment>
`;
